src/ui.rs

```diff
@@ -170,6 +233,30 @@ impl Ui {
         Ok(())
     }
 
+    pub fn finalize_writes(&self) -> io::Result<()> {
+        match &self.output_pair {
+            UiOutput::Paged { buffer } => {
+                if buffer.lock().unwrap().is_empty() {
+                    Ok(())
+                } else {
+                    let pager_cmd = pager_setting(self.settings());
+                    let mut pager = Command::new(pager_cmd)
+                        .stdin(Stdio::piped())
+                        .spawn()
+                        .expect("Failed to spawn pager");
```

Comment:
We need to start the pager earlier. If we start it only at the end, commands that can produce a lot of ouput (such as `jj log`) will need to finish producing *all* their output before the user sees any of it. In addition to high latency, it's also wasteful to produce all the output if the user just wants a few pages. FYI, the way this currently works if you do `jj log | less` (or use git's or hg's log paginated commands) is that the commands just keep producing output but the pager typically makes the writes block after a certain amount, which makes the main process also wait and not produce too much output.

Reply:
Makes sense. Then instead of writing to temporary buffer, we could spawn the pager the first time we try to write to it.

An alternative is to always spawn the pager, but as I noted in the commit message, spawning the pager but not writing to it has unintended side effects.

Reply:
> An alternative is to always spawn the pager, but as I noted in the commit message, spawning the pager but not writing to it has unintended side effects.

What was that scenario again? Can we decide that that's a broken setup of less? I think hg seems to have done that, because I'm pretty sure they spawn the pager even if they don't end up writing to it.

Reply:
> What was that scenario again?
That scenario is:

- Spawn `less`
- Write to stdout, bypassing `less`

>  Can we decide that that's a broken setup of less?
I don't think this is a broken `less` setup; this is the default setup on my machine. `LESS=X` does make the problem go away by disabling termcap init/deinit, but AIUI setting this is more of a matter of personal taste than a good default, e.g. I've seen mention that it disables features like mouse scroll support.

> I think hg seems to have done that, because I'm pretty sure they spawn the pager even if they don't end up writing to it.
Perhaps this works well for hg because once the pager is spawned, all future output is guaranteed to go to the pager because stdout/stderr get dup()-ed to the pager's stdin.

Reply:
> * Write to stdout, bypassing `less`

It seems that we just shouldn't do that. Do you mean the output that `clap` prints for us? If we don't start the pager until the command line has been successfully parsed, that shouldn't happen, I think. Or is there another case where we would do that?

Reply:
>> Write to stdout, bypassing less

> It seems that we just shouldn't do that.

If we don't ever intend to write to stdout, then isn't dup() a cleaner approach?
Or perhaps the point is that this isn't actually pager-specific; perhaps at some point we want to be able to write to arbitrary destinations in ways that might not map well to dup()-ed file descriptors.

> Do you mean the output that clap prints for us? ... Or is there another case where we would do that?

I've only observed clap doing this, and it's probably the only case where that would happen.

> If we don't start the pager until the command line has been successfully parsed, that shouldn't happen, I think.

Hm, let me try that out then.

Reply:
I think it makes a lot of sense to spawn the pager on-demand, only for specific parts of the output of specific subcommands. For example, a long-running operation which produces a lot of output at the end might want to display a progress bar directly to the terminal before switching to a pager later on.